libs/src/services/dataContracts/registryClient.js

```diff
@@ -1,20 +1,52 @@
 const Utils = require('../../utils')
+const Web3Manager = require('../web3Manager/index')
+const ProviderSelection = require('../contracts/ProviderSelection')
 
 class RegistryClient {
   constructor (web3Manager, contractABI, contractAddress) {
     this.web3Manager = web3Manager
     this.contractABI = contractABI
     this.contractAddress = contractAddress
 
-    this.web3 = this.web3Manager.getWeb3()
-    this.Registry = new this.web3.eth.Contract(contractABI, contractAddress)
+    const web3 = this.web3Manager.getWeb3()
+    this.Registry = new web3.eth.Contract(contractABI, contractAddress)
   }
 
   async getContract (contractRegistryKey) {
-    Utils.checkStrLen(contractRegistryKey, 32)
-    return this.Registry.methods.getContract(
-      Utils.utf8ToHex(contractRegistryKey)
-    ).call()
+    let providerSelector
```

Comment:
I think we could name these methods or combine them in a way that's a little more clear and also do a bit more work in the constructor

In my head we could rewrite this as
```
constructor () {
  if (web3Manager instanceof && !useExternralWeb3) {
    this.providerSelector = new ProviderSelector(endpoints)
  } else {
    this.providerSelector = null
  }
}

getContract() {
 try {
  registry.getcontract()
} catch (e) {
  if (!this.providerSelector) {
    console.error('blahblah')
    return
  }
  this.selectNewEndpoint() // this does the stuff you have lines 33 - 48
  return getContract(registryKey)
}
```

Reply:
this is a good call. i also updated the logic in RegistryClient too